### DISCUSSION TITLE
Importing Trained Weights

### DISCUSSION THREAD

**Discussion by cnrgbot** [nengo/nengo-zoo#400] [Deep Learning]
> _Archived from the Nengo Discourse forum._ Originally posted by **mvertolli** on 2017-07-14. (Forum topic #355, post #1 / id 981)

A few quick questions:

1. Is there some clear cut way to import network weights that you trained in another system? I saw this post that seems related: https://github.com/nengo/nengo_extras/issues/35

2. My understanding is that most or all Nengo ensembles are fully connected. Is there a way to do 1. with different layer types/structures (e.g., 2D convolution)?

3. Is 1. done any differently for weights of deep networks still trained in a different system?

4. Is there any convenience code kicking around for doing 1. and/or 2. with systems built in Tensorflow? I saw the work on NengoDL, but I'm less interested in Tensorflow nodes and more interested in simulating spiking versions of networks trained in Tensorflow.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **mvertolli** on 2017-07-14. (Forum topic #355, post #2 / id 982)

Okay. So, this link from another forum post answers 1. for me: https://github.com/nengo/nengo/blob/master/docs/connections.rst#direct-connections

2, 3, and 4 seem to be deeply tied to NengoDL. I literally found a convolution2D operation discussed and then removed in relation to NengoDL here: https://github.com/nengo/nengo/pull/800

I think I'm going to call this solved and just ask a bit more about NengoDL.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **mvertolli** on 2017-07-14. (Forum topic #355, post #3 / id 983)

And, here is the link that better explains how to get spiking neurons functioning with a Tensorflow setup:
http://www.nengo.ai/nengo_dl/examples/spiking_mnist.html

Don't know how I missed that.

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **Seanny123** on 2017-07-16. (Forum topic #355, post #4 / id 985)

So, have you answered your own questions now or do you require further assistance? :slight_smile:

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **bo.huang** on 2017-07-19. (Forum topic #355, post #5 / id 987)

Hello,
I also have the same problem.
In nengo_dl, there is no problem, but the question is how to do it in nengo core?
In nengo Core, if you want to do off-line learning, you use nengo.solver to solve the decoder.

            conn = nengo.Connection(input_neuron,
                                    output,
                                    synapse=None,
                                    eval_points=train_data,
                                    function=train_targets,
                                    solver=solver
                                    )

Here eval_points is your training data, function is you label. then you can solve the connection weights(decoder).
My question is, can we here first load the pre-trained weights, then the nengo solver solve the optimization problem.

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **Seanny123** on 2017-07-20. (Forum topic #355, post #7 / id 997)

Check out [this example](https://github.com/nengo/nengo_examples/blob/36a9723955073be46a118e616b83fec89a2668e1/core_examples/reuse-connection-weights.ipynb) for an explanation of how to do this in Nengo Core.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **mvertolli** on 2017-07-19. (Forum topic #355, post #6 / id 990)

For now, yes! Thanks!
